iwyu_output.cc

```diff
@@ -1062,13 +1062,43 @@ void ProcessForwardDeclare(OneUse* use,
   }
 }
 
+// Returns true if the given symbol has a mapping defined to a file, and that
+// file does not include the given FileEntry (and thus is safe to be included
+// by the FileEntry).
+static bool HasMappingSuitableForFile(
+    const string& symbol, const FileEntry* file,
+    const IwyuPreprocessorInfo* preprocessor_info) {
+  for (const std::string& public_file :
+       GlobalIncludePicker().GetCandidateHeadersForSymbol(symbol)) {
+    // To avoid cyclic includes we must only count files that do not
+    // include this one.
+    const FileEntry* public_file_entry =
+        preprocessor_info->IncludeToFileEntry(public_file);
+    if (public_file_entry == nullptr ||
```

Comment:
This nullptr check feels cowardly :) Let me see if I can understand what it means:

If `IncludeToFileEntry` returns `nullptr`, it means we haven't seen the public mapped header anywhere in our include graph, which means we cannot have included it. Is that it?

Reply:
Yes.  This happens for example in the test case for `built_ins_with_mapping-d2.h`.

Reply:
...of course, by adding a `#include` of an unseen header we could perhaps be creating a cyclic include, but I don't think there's anything that can be done about that.